src/rust/ide/lib/span-tree/src/node/kind.rs

```diff
@@ -0,0 +1,303 @@
+            _                       => false
+        }
+    }
+}
+
+
+// === API ===
+
+impl Kind {
+    /// Smart name getter.
+    pub fn name(&self) -> Option<&String> {
+        match self {
+            Self::Argument       (t) => t.name.as_ref(),
+            Self::InsertionPoint (t) => t.name.as_ref(),
+            _                        => None,
+        }
+    }
+
+    /// Smart type getter.
+    pub fn tp(&self) -> Option<&String> {
+        match self {
+            Self::Argument       (t) => t.tp.as_ref(),
+            Self::InsertionPoint (t) => t.tp.as_ref(),
+            _                        => None,
+        }
+    }
+
+    /// Smart removable flag getter.
+    pub fn removable(&self) -> bool {
+        match self {
+            Self::This     (t) => t.removable,
+            Self::Argument (t) => t.removable,
+            _                  => false,
+        }
+    }
+
+    /// Smart `ArgumentInfo` getter. Returns `None` if the node could not be attached with the
+    /// information.
+    pub fn argument_info(&self) -> Option<ArgumentInfo> {
+        match self {
+            Self::Argument       (t) => Some(ArgumentInfo::new(t.name.clone(),t.tp.clone())),
+            Self::InsertionPoint (t) => Some(ArgumentInfo::new(t.name.clone(),t.tp.clone())),
+            _                        => None
+        }
+    }
+
+    /// Smart `ArgumentInfo` setter. Returns bool indicating whether the operation was possible
+    /// or was skipped.
+    pub fn set_argument_info(&mut self, argument_info:ArgumentInfo) -> bool {
+        match self {
+            Self::Argument(t) => {
+                t.name = argument_info.name;
+                t.tp   = argument_info.tp;
+                true
+            },
+            Self::InsertionPoint(t) => {
+                t.name = argument_info.name;
+                t.tp   = argument_info.tp;
+                true
+            },
+            _ => false
+        }
+    }
+}
+
+
+// === Impls ===
+
+impl Default for Kind {
+    fn default() -> Self {
+        Self::insertion_point().into()
+    }
+}
+
+
+
+// ============
+// === This ===
+// ============
+
+/// Kind representing "this" node. For example, in the following expressions, `foo` is considered
+/// "this": `foo.bar`, `foo + bar`, `foo.+ bar`.
+#[derive(Clone,Debug,Default,Eq,PartialEq)]
+#[allow(missing_docs)]
+pub struct This {
+    pub removable : bool,
+    pub tp        : Option<String>,
+}
+
+
+// === Smart Setters ===
+
+#[allow(missing_docs)]
+impl This {
+    pub fn removable      (mut self)                    -> Self { self.removable = true     ; self }
+    pub fn typed          (mut self, tp:String)         -> Self { self.tp        = Some(tp) ; self }
+    pub fn with_removable (mut self, removable:bool)    -> Self { self.removable = removable; self }
+    pub fn with_tp        (mut self, tp:Option<String>) -> Self { self.tp        = tp       ; self }
```

Comment:
here we can name function `with_type`.

Reply:
If we leave `tp`, this should be named `with_tp`. Leaving for now.